includes/Domain/Utils/McpValidator.php

```diff
@@ -0,0 +1,370 @@
+					$errors[] = sprintf(
+						/* translators: %s: annotation field name */
+						__( 'Tool annotation field %s must be a boolean', 'mcp-adapter' ),
+						$field
+					);
+				}
+				continue;
+			}
+
+			// Validate title field.
+			if ( 'title' !== $field ) {
+				continue;
+			}
+
+			if ( ! is_string( $value ) ) {
+				$errors[] = sprintf(
+					/* translators: %s: annotation field name */
+					__( 'Tool annotation field %s must be a string', 'mcp-adapter' ),
+					$field
+				);
+				continue;
+			}
+			if ( ! empty( trim( $value ) ) ) {
+				continue;
+			}
+
+			$errors[] = sprintf(
+				/* translators: %s: annotation field name */
+				__( 'Tool annotation field %s must be a non-empty string', 'mcp-adapter' ),
+				$field
+			);
+		}
+
+		return $errors;
+	}
+
+	/**
+	 * Get validation errors for shared MCP annotations.
+	 *
+	 * Validates shared annotation fields per MCP 2025-06-18 specification:
+	 * - audience must be a non-empty array of valid Role values ("user", "assistant")
+	 * - lastModified must be a valid ISO 8601 formatted string
+	 * - priority must be a number between 0.0 and 1.0
+	 *
+	 * Only validates known shared annotation fields. Unknown fields are ignored.
+	 * Used by resources, prompts, and tools.
+	 *
+	 * @param array $annotations The annotations to validate.
+	 *
+	 * @return array Array of validation errors, empty if valid.
+	 */
+	public static function get_annotation_validation_errors( array $annotations ): array {
+		$errors = array();
+
+		foreach ( $annotations as $field => $value ) {
+			// Validate audience field.
+			if ( 'audience' === $field ) {
+				if ( ! is_array( $value ) ) {
+					$errors[] = __( 'Annotation field audience must be an array', 'mcp-adapter' );
+					continue;
+				}
+				if ( empty( $value ) ) {
+					$errors[] = __( 'Annotation field audience must be a non-empty array', 'mcp-adapter' );
+					continue;
+				}
+				if ( ! self::validate_roles_array( $value ) ) {
+					$errors[] = __( 'Annotation field audience must contain only valid roles ("user" or "assistant")', 'mcp-adapter' );
+				}
+				continue;
+			}
+
+			// Validate lastModified field.
+			if ( 'lastModified' === $field ) {
+				if ( ! is_string( $value ) || empty( trim( $value ) ) ) {
+					$errors[] = __( 'Annotation field lastModified must be a non-empty string', 'mcp-adapter' );
+					continue;
+				}
+				if ( ! self::validate_iso8601_timestamp( trim( $value ) ) ) {
+					$errors[] = __( 'Annotation field lastModified must be a valid ISO 8601 timestamp', 'mcp-adapter' );
+				}
+				continue;
+			}
+
+			// Validate priority field.
+			if ( 'priority' === $field ) {
+				if ( ! is_numeric( $value ) ) {
+					$errors[] = __( 'Annotation field priority must be a number', 'mcp-adapter' );
+					continue;
+				}
+				if ( self::validate_priority( $value ) ) {
+					continue;
+				}
+				$errors[] = __( 'Annotation field priority must be between 0.0 and 1.0', 'mcp-adapter' );
+				continue;
+			}
```

Comment:
Let's use a `switch...case` here to make our cases clear and that we're comparing against `$field`.